bytecomp/matching.ml

```diff
@@ -1259,24 +1328,38 @@ let make_constant_matching p def ctx = function
 let divide_constant ctx m =
   divide
     make_constant_matching
-    (fun c d -> const_compare c d = 0) (get_key_constant "divide")
+    (fun (_c_loc, c) (_d_loc, d) -> const_compare c d = 0)
+    (get_key_constant "divide")
     get_args_constant
     ctx m
 
 (* Matching against a constructor *)
 
 
-let make_field_args loc binding_kind arg first_pos last_pos argl =
+let make_field_args binding_kind arg first_pos last_pos field_pat_locs
+      (argl : args) =
+  let field_pat_locs = Array.of_list field_pat_locs in
+  assert (Array.length field_pat_locs = 1 + last_pos - first_pos);
   let rec make_args pos =
     if pos > last_pos
     then argl
-    else (Lprim(Pfield pos, [arg], loc), binding_kind) :: make_args (pos + 1)
+    else
+      let loc = field_pat_locs.(pos - first_pos) in
+      (Lprim(Pfield pos, [arg], loc), binding_kind, loc) :: make_args (pos + 1)
   in make_args first_pos
```

Comment:
Now that you get passed a list of the same size as the output, you should use `List.mapi` instead of `make_args`.

Reply:
I'm not actually sure I want to change this, I think the code is ok as-is.

Reply:
I am proposing to replace your version

```ocaml
let make_field_args binding_kind arg first_pos last_pos field_pat_locs
      (argl : args) =
  let field_pat_locs = Array.of_list field_pat_locs in
  assert (Array.length field_pat_locs = 1 + last_pos - first_pos);
  let rec make_args pos =
    if pos > last_pos
    then argl
    else
      let loc = field_pat_locs.(pos - first_pos) in
      (Lprim(Pfield pos, [arg], loc), binding_kind, loc) :: make_args (pos + 1)
in make_args first_pos
```

by the following

```ocaml
let make_field_args binding_kind arg first_pos last_pos field_pat_locs
      (argl : args) =
  let field_arg ofs loc =
    Lprim(Pfield (first_pos + ofs), [arg], loc, binding_kind, loc) in
  List.mapi field_arg field_pat_locs @ argl
```

Reply:
I see, I've incorporated this (and similarly in the other location).